### DISCUSSION TITLE
User OTP implementation via RESTAPI endpoints

### DISCUSSION THREAD

**Discussion by pranil420** [keycloak/keycloak#12350] [Q&A] [answered]
I fully understand the UI based integration of OTP for a user. but my concern is that i do not want to use the OTP UI provided by keycloak. I want the user to be able to integrate OTP in there account just through my application without having the hassle to switch to the keycloak UI. for this is there any api endpoint that returns us the QR generated while adding the OTP device as in the keycloak UI.

![image](https://user-images.githubusercontent.com/48387817/172109301-b5634979-3fd7-427a-b1d7-ef09141f284b.png)

as shown in the image above i want to be able to retrieve the QR data or the code that is displayed after we click on the **Unable to scan?** button. any help would be greatly appreciated.

**Comment by MariuszCwikla** [comment, marked as answer]
@pranil420 AFAIK there is no such API but it pretty straightforward to implement it on your own.
Useful links:
- https://github.com/dasniko/keycloak-extensions-demo/tree/main/rest-endpoint - this will show you how to implement custom REST endpoint
- to see how you can generate the QR code with the secret - see `org.keycloak.forms.account.freemarker.model.TotpBean.TotpBean(KeycloakSession, RealmModel, UserModel, UriBuilder)`
- to see how to do TOTP setup (user sends the one time code with the secret) - see `org.keycloak.services.resources.account.AccountFormService.processTotpUpdate()`

**Comment by pdax-maiquang** [comment]
@pranil420 did you figure out how to get the QR code from API? I'm not a java developer so really appreciate if there's any example or sample code for this functionality

**Comment by sws-lcocco** [comment]
I suggest also to Keycloak mantainer to use existing APIs to do this.

Using _reset-password_ API (`PUT /{realm}/users/{id}/reset-password` option for [user manipulation](https://www.keycloak.org/docs-api/18.0/rest-api/#_users_resource) ) a _secretValue_ can be passed as in password change:
`{ "type": "password", "userLabel": "API-set password", "value": "<password>" } `
and can be applicable to OTP registration using an appropriate OTP secret (e.g. for SHA-1 is 20-char alpahnumeric secret)
`{ "type": "totp", "userLabel": "API-set OTP", "value": "abcdefghijklm0123456" }`

I verified that secret (`value` JSON field) is written in DB into `secret_data` field corresponding to [`secretData` in Credential Model](https://github.com/keycloak/keycloak/blob/main/core/src/main/java/org/keycloak/representations/idm/CredentialRepresentation.java#L37), while `credential_data` DB field store settings for credential (e.g. algorithm, digits for OTP, etc.) corresponding to [`credentialData` in Credential Model](https://github.com/keycloak/keycloak/blob/main/core/src/main/java/org/keycloak/representations/idm/CredentialRepresentation.java#L37), and secret is passed encoded in BASE32 into OTP URL (together with user name, realm, etc.) like following template:
`otpauth://totp/<realm>:<user>?secret=<BASE32-encoded secret>&digits=6&algorithm=SHA1&issuer=<realm>&period=30`

Then, why not enable this API to set information necessary to build an OTP URL (that can be rendered as QR)?

Many thanks in advance for any suggestion, or explication of correct rationales

**Comment by Sko007** [comment]
Even this call to set otp over the reset-password endpoint. I get as response 204 But when i look into the admin ui under User credentials there is only password no otp
